rfcs/RFC-Upload-From-Dropbox.md

```diff
@@ -0,0 +1,358 @@
+
+| Code | Description                                                                                                             |
+| ---- | ----------------------------------------------------------------------------------------------------------------------- |
+| 401  | if dataset_id or submission_id does not exist, or if the user does not own the submission or upload in-progress upload. |
+| 400  | if the parameters supplied are invalid.                                                                                 |
+
+#### Upload Status and State
+
+The API documentation returns a _status_ in response. This status informs the user of the current state of the upload.
+An upload follows this state diagram:
+
+![Upload States](https://app.lucidchart.com/publicSegments/view/5fe0858f-eb93-4b64-9842-6f25cad371f3/image.png)
+Figure 2: Upload State Diagram
+
+There are several different states that an upload can be in. The table below defines them.
+
+| #   | states         | Description                                                                                 |
+| --- | -------------- | ------------------------------------------------------------------------------------------- |
+| S0  | Start          | The upload is never actually in this state.                                                 |
+| S1  | Waiting        | The upload is enqueued, and waiting for computing resources.                                |
+| S2  | uploading      | The file is actively being uploaded.                                                        |
+| S3  | Completed      | The upload was completed successfully.                                                      |
+| S4  | Failed         | The upload has failed. A new link or file must be uploaded. Any upload progress is deleted. |
+| S5  | Cancel Pending | The upload is in the process of being canceled                                              |
+| S6  | Canceled       | The upload has been canceled. Any upload progress is deleted.                               |
+
+These states will be stores in the upload table until the upload is complete.
+
+##### S0 - >S1
+
+An upload moves from S0 to S1 when an upload request has been made, and the system is waiting for resources to begin uploading.
+
+##### S1 - >S2
+
+An upload moves from S1 to S2 when the file is actively being uploaded to S3.
+
+##### S2 -> S5
+
+Once a user requests an upload to be canceled the state changes to Cancel Pending. The cancellation may not happen
+immediately, therefore the status of the upload is changed to cancel pending until the upload has been canceled by the
+compute resource processing the job.
+
+##### S5 -> S6
+
+Once the upload has been canceled by the compute resource the state will be canceled.
+
+##### S2 -> S3
+
+The upload has been completed. The state of the upload is complete.
+
+##### S2 -> S1
+
+If the upload fails, the job returns to the queue to be retried at a later time. The state changes to waiting.
```

Comment:
We need to keep track of failed attempts somewhere (likely the same dynamoDB instance) to know when to give up and got into error.

Reply:
SQS will track the number of times the download has been attempted and will put the failed download into the dead letter queue. This does not track the actual error that was thrown, but we can log something in cloudwatch from whatever compute resource we use.